### PR TITLE
[WIP] remote: cache_exists performance

### LINKED ISSUE #1762: perf: remote: status: don't pull full list of files from remote
Currently, we retrieve the full list of files from remote and then compare it with local cache, which gets really slow if user has lots of files on the remote already. So instead of doing this https://github.com/iterative/dvc/blob/0.32.1/dvc/remote/local.py#L591 we should compare it with local cache more carefully and efficiently. 

https://discordapp.com/channels/485586884165107732/485596304961962003/558288115249184775

### PR DESCRIPTION
**Pull request by ei-grad**
This pull-request implements an ability to check the remote files existence with iteration over local cache files names instead of fetching the whole remote files list. It adds the `no_traversal` config option to the remote config block, which is needed to be switched on to enable this behavior.

It also adds the progress bar which display:
- case `no_traversal = false` (original behavior): the amount of files to be checked which have been found during the iteration over the `remote.all()` remote files list
- case `no_traversal = true`: count of checked files

TODO:
* [ ] recheck and complete the implementations for every remote type
* [ ] write tests for cache_exists

Fix #1762

### CODE REVIEW

File: dvc/config.py [hunk 1]
```diff
@@ -132,14 +132,16 @@ class Config(object):  # pylint: disable=too-many-instance-attributes
     CONFIG = "config"
     CONFIG_LOCAL = "config.local"
 
+    BOOL = And(str, is_bool, Use(to_bool))
```
Comment (efiop) [inline dvc/config.py]: ❤️

File: dvc/output/base.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def __init__(
         self.repo = stage.repo
         self.url = path
         self.info = info
-        self.remote = remote or self.REMOTE(self.repo, {})
+        self.remote = remote or self.REMOTE(self.repo, {}, "output")
```
Comment (shcheklein) [inline dvc/output/base.py]: does not seem that `output` is the right name here. Should be it just cache in this case?

File: dvc/remote/base.py [hunk 3]
```diff
@@ -288,16 +294,71 @@ def changed_cache(self, checksum):
...
+        threads in parallel
+        """
+        results = []
+        args = threadsafe_iterator(args)
+        with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
+            futures = [executor.submit(func, args) for i in range(self.JOBS)]
+            for future in concurrent.futures.as_completed(futures):
+                try:
+                    data = future.result()
+                except Exception:
```
Comment (shcheklein) [inline dvc/remote/base.py]: seems not the right way to handle this. What kind of errors do we except? Could we actually ignore them and return a meaningful result?

File: dvc/remote/base.py [hunk 4]
```diff
@@ -288,16 +294,71 @@ def changed_cache(self, checksum):
...
         return False
 
+    def _parallel(self, func, args):
+        """Wraps iterator into threadsafe_iterator and executes func in JOBS
+        threads in parallel
+        """
+        results = []
+        args = threadsafe_iterator(args)
+        with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
+            futures = [executor.submit(func, args) for i in range(self.JOBS)]
```
Comment (shcheklein) [inline dvc/remote/base.py]: I'm trying to understand this `threadsafe_iterator` and the way it's used with `ThreadPoolExecutor`. Is it to save some memory? It feels like something like queue could be used or we just can submit individual items, no?

File: dvc/remote/base.py (hunk 3 above)
Comment (shcheklein) [reply dvc/remote/base.py]: especially considering that logic with iterator - we'll be ignoring a big chuck of the result potentially

File: dvc/remote/base.py (hunk 4 above)
Comment (shcheklein) [reply dvc/remote/base.py]: Or is it just because of that progess loop wrapper? Btw, could we just reuse the progress_callback mechanism here?

Comment (ei-grad) [reply dvc/remote/base.py]: > It feels like something like queue could be used or we just can submit individual items, no?
Yes, it transforms the iterator into the queue, exactly.

**Pull request closed without merging**